README
# Copyright (C) 2013-2019 Wilbert van Ham, Radboud University Nijmegen
# Distributed under the terms of the GNU General Public License (GPL).

RuSocSci is a package 
for connecting to the Radboud University Nijmegen, Faculty of Social 
Sciences USB Buttonbox and Red Joystick in a PsychoPy-like way.

Starting with version 0.8 this package is both Python 2 and Python 3 compatible.


CHANGES
0.1 first version
0.2 corrected wrong url
0.3 added OSX support, now supports Linux, Windows and OSX. note that for window administrative privilieges are needed
0.4 led output support
0.5 joystick support
0.6 buttonbox is now an object, api change
0.6.4 common code for buttonbox and joystick, optional port for both
0.6.5 support for ACM chip devices (Uno)
0.6.6 added HideStdout
0.6.7 added BITSI extended (extended.py)
0.6.8 detect port bug in windows, new url
0.7.0 added waitKeys/getKeys link to waitButtons/getButtons, return psychopy style for waitButtons (None/['A'])
0.7.2 adopted init to See "Automatic (Software) Reset" in the documentation.
0.7.6 errors in case of id string timeout of extended are now info's, complies with faulty buttonboxes
0.7.7 compatible with Python 3 and new style buttonbox id strings with version number
0.7.8 bugfix in open
0.7.9 bugfix for python3 in joystick
0.8 fully python 3 compatible
0.8.1 no changes
0.8.2 bug in sendMarker/setLeds of buttonbox
0.8.3 bug in send of extended
0.8.4 remove debug info
0.8.5 move to github, 2 bugs in python 3 MS
0.8.6 add waitButtonsHog
0.8.7 read id string immediately after connecting for devices that do not reset (5 button buttonbox)
      increased connect timeout to 6 s
      do not set timeout of device if already 0s (prevent error in windows)
      pylint code
0.9   add sleep function that prevents windows hang detect in msw with waitButtons
0.9.1 remove buttonbox2.py


rusocsci/extended.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""RuSocSci module for the BITSI extended buttonbox

Copyright (C) 2013-2019 Wilbert van Ham, Radboud University Nijmegen
Distributed under the terms of the GNU General Public License (GPL) version 3 or newer.
"""
import sys
import time
import logging
from . import utils, buttonbox # p23

class Extended(buttonbox.Buttonbox):
	"""
	Object that connects tot RuSocSci buttonbox in extended mode. Typical usage::

		from rusocsci import extended

		bb = extended.Extended() # connect to last inserted buttonbox

		print("Press any key")
		l = bb.waitKeys(10)
		if l:
			print("key pressed: {}".format(l[0]))
		else:
			print("no key pressed.")

	or for non-blocking input (using the Python 3 print function)::

		import time, sys
		from rusocsci import buttonbox

		bb = buttonbox.Buttonbox() # connect to last inserted buttonbox

		print("Press any key")
		while not len(bb.getKeys()): # while there is no input
			print(".", end='')
			sys.stdout.flush()
			time.sleep(1)

	"""
	def __init__(self, id=0, port=None):
		self._device, idString = utils.open(utils.getPort(id, port))
		idStringCorrect = "BITSI_extend mode, Ready!"
		if not self._device:
			logging.error("No BITSI extended buttonbox connected.")
		elif idString[:len(idStringCorrect)] == idStringCorrect:
			logging.debug("Device is a BITSI extended buttonbox: %s", idString)
		else:
			logging.info("Device did not identify as a BITSI extended buttonbox: %s", idString)
		self.calibratedSound = False
		self.calibratedVoice = False

	def send(self, val):
		"""
		Set buttonbox LEDs to a certain pattern
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")
		#self._device.write(bytes((val,))) # p23

		if sys.version_info >= (3, 0): #0.8.3
			self._device.write(bytes([val]))
		else:
			self._device.write(chr(val))

	def setLeds(self, leds=None, val=None):
		"""
		connect leds to signals output by computer
		"""
		if leds is None:
			leds = [False, False, False, False, False, False, False, False]
		if val is None:
			val = 0
			for i in range(8):
				if len(leds) > i:
					if leds[i]:
						val += 1<<i
				else:
					break
		self.send(ord('L'))
		self.send(ord('O'))
		self.send(val)

	def sendMarker(self, markers=None, val=None):
		"""
		send marker to buttonbox output
		"""
		if markers is None:
			markers = [False, False, False, False, False, False, False, False]
		if val is None:
			val = 0
			for i in range(8):
				if len(markers) > i:
					if markers[i]:
						val += 1<<i
				else:
					break
		self.send(ord('M'))
		self.send(val)

	def calibrateSound(self):
		"""
		define silence as the volume in the next second
		"""
		self.send(ord('C')) # calibrate sound
		self.send(ord('S'))
		time.sleep(1) # make sure to be silent during the calibration
		self.calibratedSound = True

	def waitSound(self, maxWait=float("inf"), flush=True, buttonList='S'):
		"""
		wait till sound is detected
		"""
		if not self.calibratedSound:
			logging.debug("calibrating sound, wait 1 s.")
			self.calibrateSound()
		if flush:
			self.clearEvents() # flush possible awaiting sound events
		self.send(ord('D')) # detect sound
		self.send(ord('S'))
		return self.waitButtons(buttonList=buttonList, flush=False, maxWait=maxWait)

	def calibrateVoice(self):
		"""
		define silence as the volume in the next second
		"""
		self.send(ord('C')) # calibrate
		self.send(ord('V'))
		time.sleep(1) # make sure to be silent during the calibration
		self.calibratedVoice = True

	def waitVoice(self, maxWait=float("inf"), flush=True, buttonList='V'):
		"""
		wait till sound is detected
		"""
		if not self.calibratedVoice:
			logging.debug("calibrating voice, wait 1 s.")
			self.calibrateVoice()
		if flush:
			self.clearEvents() # flush possible awaiting sound events
		self.send(ord('D')) # detect sound
		self.send(ord('V'))
		return self.waitButtons(buttonList=buttonList, flush=False, maxWait=maxWait)


rusocsci/joystick.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""RuSocSci module for the red joystick

Copyright (C) 2013-2019 Wilbert van Ham, Radboud University Nijmegen
Distributed under the terms of the GNU General Public License (GPL) version 3 or newer.

Known issues:
    - All usb2serial devices are detected. The list of joysticks therefore also contains
      buttonboxes.

"""
import sys, serial, time, os, re, logging, struct
from . import utils

# our buttonbox and joystick have id 0403:6001 from its UART IC
# make sure you have the pyusb module installed
# for MS Windows you may need this: http://sourceforge.net/apps/trac/libusb-win32/wiki
def getNumJoysticks():
	""""
	Return a count of the number of joysticks available. 
	This really returns a count of all usb to serial devices
	"""
	return len(utils.serialList())

class Joystick():
	"""
	Typical usage::

		from rusocsci import joystick
		import time

		nJoys = joystick.getNumJoysticks() # to check if we have any
		id=0
		joy = joystick.Joystick(id)#id must be <= nJoys-1

		while True:#while presenting stimuli
			x = joy.getX()
			time.sleep(1)
	"""
	def __init__(self, id=0, port=None):
		self.x = 126 # angle range is [51..201]
		[self._device, idString] = utils.open(utils.getPort(id, port))

		if idString == "joystick streaming angle, Ready!":
			logging.debug("Device is a joystick ({}): {}".format(len(idString), idString))
		else:
			logging.error("Device is NOT a joystick ({}): {}".format(len(idString), idString))

	def __del__(self):
		if self._device:
			self._device.close()

	def getNumHats(self):
		"Get the number of hats on this joystick"
		return 0
	def getX(self):
		"Returns the value on the X axis (equivalent to joystick.getAxis(0))"
		if self._device == None:
			logging.error("Joystick not initialized")
			return -1
		self._device.timeout = 0
		c = self._device.read()
		nBytes = 0
		while len(c) != 0:
			self.x = struct.unpack('B', c)[0]
			nBytes+=1
			c = self._device.read()
		logging.debug("read {} bytes, ending in {}".format(nBytes, self.x))
		return self.x

	def getAllAxes(self):
		"""
		Get a list of all current axis values
		"""
		return [self.getX()]
	def getNumAxes(self):
		"""
		Returns the number of joystick axes found
		"""
		return 1
	def getAxis(self, axisId):
		"""
		Get the value of an axis by an integer id (from 0 to number of axes-1)
		"""
		return self.getX()
	
	def clearEvents():
		if _device == None:
			event.flushInput(keyboard)
		else:
			self._device.flushInput()
		


rusocsci/buttonbox.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
RuSocSci module for the BITSI buttonbox

Copyright (C) 2013-2019 Wilbert van Ham, Radboud University Nijmegen
Distributed under the terms of the GNU General Public License (GPL) version 3 or newer.
"""
import sys
import time
import logging
from . import utils # p23

# our buttonbox has id 0403:6001 from its UART IC
# todo: add support for 2341:0001 (Arduino Uno)
# make sure you have the pyusb module installed
# for window you may need this: http://sourceforge.net/apps/trac/libusb-win32/wiki

class Buttonbox(object):
	"""
	Object that connects tot RuSocSci BITSI buttonbox. Typical usage::

		from rusocsci import buttonbox

		bb = buttonbox.Buttonbox() # connect to last inserted buttonbox

		print("Press any key")
		l = bb.waitKeys(10)
		if l:
			print("key pressed: {}".format(l[0]))
		else:
			print("no key pressed.")

	or for non-blocking input (using the Python 3 print function)::

		from __future__ import print_function
		from rusocsci import buttonbox
		import time, sys

		bb = buttonbox.Buttonbox() # connect to last inserted buttonbox

		print("Press any key")
		while not len(bb.getKeys()): # while there is no input
			print(".", end='')
			sys.stdout.flush()
			time.sleep(1)

	"""
	def __init__(self, id=0, port=None):
		self._port = utils.getPort(id, port)
		self._device = None
		self._maxWait = None # overrides maxWait in waitButtons
		if not self._port:
			raise Exception("No USB serial device connected, could not get port name.")
		self._device, idString = utils.open(self._port)
		# idString may contain the rrormessage, for instance "USB serial timeout"
		if not self._device:
			logging.critical("No BITSI buttonbox connected, could not connect to port %s", self._port)
		elif idString.startswith("BITSI mode, Ready!") or idString.startswith("BITSI event mode, Ready!"):
			logging.debug("Device is a BITSI buttonbox (%d): %s", len(idString), idString)
		else:
			logging.error("Device is NOT a BITSI buttonbox (%d): %s", len(idString), idString)

	def close(self):
		"""
		Close serial connection to buttonbox
		"""
		if self._device:
			self._device.close()

	def clearEvents(self):
		"""
		Clear previous events that are still in the input buffer.
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")
		try:
			self._device.flushInput()
		except Exception as e:
			raise Exception("Could not clear buttonbox buffer:\n{}".format(e))

	def getButtons(self, buttonList=None):
		"""
		Returns a list of buttons that were pressed on the buttonbox.

		:Parameters:
			buttonList : **None** or []
				Allows the user to specify a set of keys to check for.
				Only keypresses from this set of keys will be removed from the keyboard buffer.
				If the keyList is None all keys will be checked and the key buffer will be cleared
				completely. NB, pygame doesn't return timestamps (they are always 0)
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")

		if self._device.timeout != 0:
			self._device.timeout = 0
		if sys.version_info >= (3, 0):
			cList = self._device.read(1024).decode() # p23, warning utf8, expect trouble for non-ascii bytes
			# <class 'str'>
		else:
			cList = self._device.read(1024)
			# <type 'str'>
		cListSelected = []
		for c in cList:
			if buttonList is None or c in buttonList:
				cListSelected.append(c)
		return cListSelected

	def getKeys(self, keyList=None):
		"""
		Mutatis Mutandis identical to getButtons
		"""
		return self.getButtons(buttonList=keyList)

	def waitButtons(self, maxWait=float("inf"), buttonList=None, timeStamped=False, flush=True):
		"""
		Same as getButtons(), but halts everything (including drawing) while awaiting
		input from buttonbox. Implicitly clears buttonbox, so any preceding buttonpresses will be lost.

		:Parameters:
			maxWait : any numeric value.
				Maximum number of seconds period and which buttons to wait for.
				Default is float('inf') which simply waits forever.
			buttonList:
				List of one character strings containing the buttons to react to.
				All other button presses will be ignored. Note that for BITSI
				buttonboxes the buttons are identified by capital letters upon press
				and by lower case letters upon release.

		Returns None if times out. Returns a list of one character upon succes, like
		the PsychoPy event module.
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")

		if flush:
			self._device.flushInput()
		t = time.time()
		while maxWait > time.time() - t:
			if maxWait == float("inf"):
				self._device.timeout = None
			else:
				self._device.timeout = maxWait - (time.time() - t)
			c = self._device.read(1).decode()
			if buttonList is None or c in buttonList:
				# return
				break
			else:
				# as if nothing pressed
				c = ''
		if c == '':
			# version 0.7 addition to comply with PsychoPy
			return None
		if hasattr(timeStamped, 'timeAtLastReset'):
			return [(c, time.time() - timeStamped.timeAtLastReset)]
		elif timeStamped:
			# return as a one item list to mimic getButtons behaviour
			return [(c, time.time() - t)]
		return [c] # version 0.7 change to comply with PsychoPy

	def waitButtonsHog(self, maxWait=float("inf"), buttonList=None, timeStamped=False, flush=True):
		"""
		Same as waitButtons(), but using processor hogging rather than an interrupt.
		May start faster than waitButtons in MS Windows if maxWait is given.

		:Parameters:
			maxWait : any numeric value.
				Maximum number of seconds period and which buttons to wait for.
				Default is float('inf') which simply waits forever.
			buttonList:
				List of one character strings containing the buttons to react to.
				All other button presses will be ignored. Note that for BITSI
				buttonboxes the buttons are identified by capital letters upon press
				and by lower case letters upon release.

		Returns None if times out. Returns a list of one character upon succes, like
		the PsychoPy event module.
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")
		t = time.time()

		if flush:
			self._device.flushInput()
		cnt = self._device.inWaiting()
		c = ''
		while (maxWait > time.time() - t) and (c == ''):
			#  The following timeout setters cause an unwanted delay
			#if maxWait == float("inf"):
			#             self._device.timeout = None
			#else:
			#             self._device.timeout = maxWait - (time.time() - t)
			if cnt != self._device.inWaiting():
				s = self._device.read(self._device.inWaiting())
				s = s[cnt:len(s)]
				cnt = self._device.inWaiting()
				i = 0
				# get the first response after this function was called.
				while i < len(s) and c == '':
					if buttonList is None or s[i] in buttonList:
						c = s[i].decode()
						i += 1

		if c == '':
			# version 0.7 addition to comply with PsychoPy
			return None

		if hasattr(timeStamped, 'timeAtLastReset'):
			return [(c, time.time() - timeStamped.timeAtLastReset)]
		elif timeStamped:
			# return as a one item list to mimic getButtons behaviour
			return [(c, time.time() - t)]
		return [c] # version 0.7 change to comply with PsychoPy

	def waitKeys(self, maxWait=float("inf"), keyList=None, timeStamped=False):
		"""
		Mutatis Mutandis identical to waitButtons
		"""
		return self.waitButtons(maxWait=maxWait, buttonList=keyList, timeStamped=timeStamped, flush=True)

	def setLeds(self, leds=None, val=None):
		"""
		Set buttonbox LEDs to a certain pattern
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")
		if leds is None:
			leds = [False, False, False, False, False, False, False, False]
		if val is None:
			val = 0
			for i in range(8):
				if len(leds) > i:
					if leds[i]:
						val += 1<<i
				else:
					break
		#self._device.write(chr(val))
		#self._device.write(bytes([val])) #p23, list instead of tuple in 0.8.2
		if sys.version_info >= (3, 0): #0.8.2
			self._device.write(bytes([val]))
		else:
			self._device.write(chr(val))

	def sendMarker(self, markers=None, val=None):
		"""
		Mutatis Mutandis identical to setLeds()
		"""
		return self.setLeds(leds=markers, val=val)

	def waitLeds(self, leds=None, wait=1.0, val=None):
		"""
		Set buttonbox LEDs to a certain pattern and wait a while. Reset afterwards.
		"""
		if self._device is None:
			raise Exception("No buttonbox connected")
		self.setLeds(leds, val)
		time.sleep(wait)
		self.setLeds()


rusocsci/utils.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""RuSocSci module for utilities, like listing USB to serial devices and connecting to them.

# Copyright (C) 2013-2019 Wilbert van Ham, Radboud University Nijmegen
# Distributed under the terms of the GNU General Public License (GPL) version 3 or newer.

Known issues:
    - All usb to serial devices are detected. The list of devices therefore contains
      joysticks, buttonboxes and other usb to serial devices.

"""
import sys, serial, time, os, re, logging, struct

TIMEOUT = 6.0
if sys.platform == "win32":
	if sys.version_info >= (3, 0):
		import winreg as _winreg
	else:
		import _winreg

class HideStderr(object):
	'''
	A context manager that blocks stderr for its scope, usage::

		from rusocsci import utils
		import os
		with utils.HideStderr():
			os.system('ls noexistentfile') # error will not show

	'''

	def __init__(self, *args, **kw):
		sys.stderr.flush()
		self._origstderr = sys.stderr
		self._oldstderr_fno = os.dup(sys.stderr.fileno())
		self._devnull = os.open(os.devnull, os.O_WRONLY)

	def __enter__(self):
		self._newstderr = os.dup(2)
		os.dup2(self._devnull, 2)
		os.close(self._devnull)
		sys.stderr = os.fdopen(self._newstderr, 'w')

	def __exit__(self, exc_type, exc_val, exc_tb):
		sys.stderr = self._origstderr
		sys.stderr.flush()
		os.dup2(self._oldstderr_fno, 2)
		os.close(self._oldstderr_fno)
		
class HideStdout(object):
	'''
	A context manager that blocks stdout for its scope, usage::

		from rusocsci import utils
		import os
		with utils.HideStdout():
			os.system('ls -l') # output will not show
	'''

	def __init__(self, *args, **kw):
		sys.stdout.flush()
		self._origstdout = sys.stdout
		self._oldstdout_fno = os.dup(sys.stdout.fileno())
		self._devnull = os.open(os.devnull, os.O_WRONLY)

	def __enter__(self):
		self._newstdout = os.dup(1)
		os.dup2(self._devnull, 1)
		os.close(self._devnull)
		sys.stdout = os.fdopen(self._newstdout, 'w')

	def __exit__(self, exc_type, exc_val, exc_tb):
		sys.stdout = self._origstdout
		sys.stdout.flush()
		os.dup2(self._oldstdout_fno, 1)
		os.close(self._oldstdout_fno)


# our buttonbox and joystick have id 0403:6001 from its UART IC
# make sure you have the pyusb module installed
# for MS Windows you may need this: http://sourceforge.net/apps/trac/libusb-win32/wiki
# FTDIBUS
def _winList(service, bus=None, suffix=None):
	""""
	Helper for windows for serialList function
	"""
	try:
		# find number of FTDI USB serial devices attached
		reg = _winreg.ConnectRegistry(None, _winreg.HKEY_LOCAL_MACHINE)
		keyString = r"SYSTEM\CurrentControlSet\services""\\" + service + r"\Enum"
		key = _winreg.CreateKeyEx(reg, keyString, 0, _winreg.KEY_READ)
		count = _winreg.QueryValueEx(key, "Count")[0]
	except Exception as e:
		logging.info("Could not create service key string: {}".format(keyString))
		return []
	
	if count == 0:
		logging.debug("No "+ service +" USB serial device connected.")
	if count > 1:
		logging.debug("Multiple "+ service +"  USB serial connected, last connected is first in list.")
		
	devices = []
	for i in range(count-1, -1, -1):
		value = _winreg.QueryValueEx(key, str(i))[0]
		devices.append(value)
	logging.debug("USB serial devices ({}): {}".format(service, devices))
	
	# find the corresponding serial ports
	ports = []
	for device in devices:
		#device is something like: ur"USB\VID_0403&PID_6001\AH01L3OL" or "USB\VID_2341&PID_0043\6493433313535141D0B2"
		id = device[22:]
		# find port, the following line gives an error in windows without administrative priviliges
		try:
			if bus:
				keyString = r"SYSTEM\CurrentControlSet\Enum""\\" + bus + id + suffix + r"\Device Parameters"
			else:
				keyString = r"SYSTEM\CurrentControlSet\Enum""\\" + device + r"\Device Parameters"
			key = _winreg.CreateKey(reg, keyString)
		except Exception as e:
			logging.info("Could not create device key string: {}".format(keyString))
			return []
		try:
			value = _winreg.QueryValueEx(key, "PortName")[0]
			ports.append(value)
		except Exception as e:
			logging.info("Could not find USB serial device, no value for PortName: {}\n".format(e))
	return [s.encode('ascii', 'ignore') for s in ports]

def serialList():
	"""
	Get a list of USB to serial devices connected. The returned list is that of the serial ports.
	"""
	if sys.platform == "win32":
		ports = _winList("FTDIBUS", r"FTDIBUS\VID_0403+PID_6001+", r"A\0000") # FTDI based (Arduino 2009)
		ports += _winList("FTDIBUS", r"FTDIBUS\VID_0403+PID_6001+", r"\0000") # sometimes this works
		ports += _winList("usbser") # ACM based, Arduino Uno, Uno R3, Leonardo
		# usbser should be generic and work for all USB to serial interfaces
		logging.debug("USB serial ports: {}".format(ports))
		return ports
	elif sys.platform == "darwin":
		# /dev/tty.usbserial* for FTDI (0403:6001) in Arduino 2009
		# /dev/tty.usbmodem* for Arduino Uno R3 (2341:0043)
		# the latter probably works for Arduino Leonardo (2341:8036) as well (todo: test)
		with HideStderr():
			s = os.popen('ls -t /dev/tty.usbserial* /dev/tty.usbmodem*').read().rstrip()
		ports = s.split()
		if len(ports)==0:
			logging.info("Could not find USB serial device. Install the FTDI VCP/Arduino driver and plug in the device.")
		if len(ports) > 1:
			logging.debug("Multiple USB serial devices connected, last connected is first in list.")
		logging.debug("USB to serial interfaces: {}".format(ports))
		return ports
	else:
		# /dev/serial/by-id/*FTDI* for FTDI (0403:6001) in Arduino 2009
		# /dev/ttyACM? for Arduino Uno R3 (2341:0043) and Arduino Leonardo (2341:8036)
		# /dev/ttyACM? for Teensy
		try:
			# HideStderr will fail in openSesame since it has no stderr
			with HideStderr():
				s = os.popen('ls -t /dev/serial/by-id/*FTDI* /dev/ttyACM?').read()
		except:
			s = os.popen('ls -t /dev/serial/by-id/*FTDI* /dev/ttyACM?').read()
		ports = s.split()
		if len(ports)==0:
			logging.info("Could not find USB serial device.")
		if len(ports) > 1:
			logging.debug("Multiple USB serial devices connected, last connected is first in list.")
		logging.debug("USB to serial interfaces: {}".format(ports))
		return ports


def getPort(id=0, port=None):
	""" 
	Return serial port.
	id is position in list (0 is latest), port is 'COM1', /dev/tty, 0 or something similar.
	Note that port can also be a real serial port (non USB).
	returns serial connection and id string. 
	"""
	if port != None:
		return port
	else:
		ports = serialList()
		if len(ports)>id:
			port = ports[id]
		else:
			logging.error("id {} invalid, number of USB serial ports found: {}".format(id, len(ports)))
			return None
	if port == None or port == "":
		logging.error("No USB serial port found.")
		return None
	else:
		logging.debug("Connecting to USB serial port: {}".format(port))
	return port

def open(port):
	""" 
	Open serial port.
	port is 'COM1', /dev/tty, 0 or something similar.
	Note that port can also be a real serial port (non USB).
	returns serial connection and id string. In Python 3 this is a string object.
	"""
	
	# port must be str in Python 3, str or unicode in Python 2: convert
	# Python2: converting to unicode is harmless, open() accepts both unicode and str
	# Python3: converting from bytes to unicode is necessary, converting from unicode to unicode will fail
	try:
		port = port.decode("utf-8")
	except:
		pass


	device = None
	while True:
		try:
			device = serial.Serial(port, baudrate=115200, parity='N', timeout = 0.0)  # open serial port
#		except Exception as e:
		except serial.serialutil.SerialException as e:
			if 'Device or resource busy:' in e.__str__():
				logging.debug('Opening COM port is taking a little while, please stand by...')
			else:
				logging.error("Could not connect to serial port {}: \n{}".format(port, e))
				return [None, "Could not connect to serial port"]
		if device is not None:
			break
		time.sleep(1)

		
	# reset connection
	device.flushInput() # race condition, must happen before device sends ID string for devices that do not reset, like Teensy
	# The following three lines cause a reset of the Arduino. 
	# See "Automatic (Software) Reset" in the documentation.
	# Make sure not to send data for the first second after reset.
	device.setDTR(False)
	time.sleep(0.1)
	device.setDTR(True)
	time.sleep(1.0)
	
	#idString = "joystick streaming angle, Ready!" + "0d0a".decode("hex")
	# collect byes up to "!\x0d\x0a" that identify the type of device
	beginTime = time.time()
	bytesRead = b"" # p23
	# read till windows newline for a maximum of 3 seconds
	while len(bytesRead) < 2 or bytesRead[-2:] != b"\x0d\x0a": # p23
		if time.time() > beginTime + TIMEOUT:
			logging.info("USB serial timeout waiting for ID string, got: '{}'".format(bytesRead))
			return (device, "USB serial timeout")
		bytesReading = device.read() # p2: <type 'str'>, p3: <class 'bytes'> (immutable)
		bytesRead += bytesReading
		
	return (device, bytesRead[:-2].decode())

def sleep(duration, singleDuration=1.0, sleepFunc=time.sleep, timeFunc=time.time, func=None):
	"""
	wait while doing something periodically
	example use:
		sleep(20) # does the same as time.sleep
		sleep(20, func=event.clearEvents # prevents windows from detecting hang
		sleep(20, timeFunc=time.perf_counter, func=event.clearEvents # same as above
		sleep(20, sleepFunc=event.waitKeys, func=event.clearEvents # exit on keypress
	"""
	t0 = timeFunc()
	while True:
		t1 = timeFunc()-t0 # time since start
		durationLeft = duration - t1 
		if durationLeft <= singleDuration: # just the normal version
			return sleepFunc(durationLeft)
		else:
			retval = sleepFunc(singleDuration)
			if retval is not None:
				return retval
			if func is not None:
				func()


